src/batched/dense/impl/KokkosBatched_Gemm_Armpl_Impl.hpp

```diff
@@ -0,0 +1,300 @@
+    __Bistrd = __Bjstrd * __Bn;
+    __Bbstrd = __Bistrd * __Bm;
+
+    __Cjstrd = __ninter;
+    __Cistrd = __Cjstrd * __Cn;
+    __Cbstrd = __Cistrd * __Cm;
+
+    __transa = std::is_same<ArgTransA, Trans::NoTranspose>::value ? 'N' : 'T';
+    __transb = std::is_same<ArgTransB, Trans::NoTranspose>::value ? 'N' : 'T';
+  }
+
+  int invoke() {
+    if (__handle->enableDebug) {
+      std::cerr << "__nbatch:" << std::to_string(__nbatch)
+                << ", __ninter:" << std::to_string(__ninter)
+                << ", __Am:" << std::to_string(__Am)
+                << ", __An:" << std::to_string(__An)
+                << ", __alpha:" << std::to_string(__alpha)
+                << ", __beta:" << std::to_string(__beta) << std::endl;
+    }
+
+    if (!std::is_same<avt, double>::value ||
+        !std::is_same<bvt, double>::value ||
+        !std::is_same<cvt, double>::value ||
+        !std::is_same<ScalarType, double>::value) {
+      std::ostringstream os;
+      os << "KokkosBatched::Impl::BatchedArmplGemm only supports 'double' "
+            "scalar types."
+         << std::endl;
+      Kokkos::Impl::throw_runtime_exception(os.str());
+    }
+
+    if (__nbatch != 0) {
+      if (__ninter == 0 || __nbatch % __ninter) {
+        std::string msg =
+            "batch size must be evenly divisible by ninter. __nbatch: ";
+        msg += (std::to_string(__nbatch) +
+                ", __ninter: " + std::to_string(__ninter) + "\n");
+        Kokkos::abort(msg.c_str());
+      }
+
+      // Calculate internal batch size for interleaving
+      __nbatch /= __ninter;
+
+      // Allocate space for interleaving
+      //   __Adp and __Bdp are deleted in __run()
+      //   __Cdp is deleted in __repack_view()
+      __Adp = new avt[__Abstrd * __nbatch];
+      __Bdp = new bvt[__Bbstrd * __nbatch];
+      __Cdp = new cvt[__Cbstrd * __nbatch];
+
+      __unpack_views();
+      __run();
+      __repack_view();
+    }
+    return 0;
+  }
+};
+/********************* END non-functor-level routines *********************/
+}  // namespace Impl
+}  // namespace KokkosBatched
+#else   // KOKKOSKERNELS_ENABLE_TPL_ARMPL
+namespace KokkosBatched {
+namespace Impl {
+/********************* BEGIN non-functor-level routines *********************/
+// TODO: wrap this class in a macro for permutations of supported scalars.
+template <class ArgTransA, class ArgTransB, class ArgBatchSzDim,
+          class HandleType, class ScalarType, class AViewType, class BViewType,
+          class CViewType>
+class BatchedArmplGemm {
+ public:
+  BatchedArmplGemm(HandleType *const handle, ScalarType alpha, AViewType A,
+                   BViewType B, ScalarType beta, CViewType C) {
+    (void)handle;
+    (void)alpha;
+    (void)A;
+    (void)B;
+    (void)beta;
+    (void)C;
+  }
+
+  int invoke() {
+    std::ostringstream os;
+    os << "KokkosBatched::Impl::BatchedArmplGemm requires the ARMPL TPL"
+       << std::endl;
+    Kokkos::Impl::throw_runtime_exception(os.str());
```

Comment:
@e10harvey  I am fine with throwing exception here but I suggest throwing exception earlier, either in ```KokkosBatched_Gemm_Decl.hpp``` or when creating handle if KOKKOSKERNELS_ENABLE_TPL_ARMPL is not defined.

Reply:
We could potentially throw in the Decl file but I'd like to avoid handling special cases for different algorithm types in that file for clarity in the long term. I'd actually prefer to just have a table of function pointers indexed on the algorithm type rather than the switch statement we have. Even for the heuristic cases, I would like the conditions to be handled in an Impl file rather than cluttering the Decl file with those conditions. @vqd8a: Are you open to moving forward in this direction?

I do not think we can throw when the handle is created since we do not have information about user calls at that point.

Reply:
I agree that it should not be in the Decl file for clarity purpose.
I thought we know whether a user wants to use ARMPL or not at the time the handle is created. If this is impossible, I am okay with the way it is now.

Reply:
You're right. The algo type is protected so the only way to set it is via the constructor; I will update accordingly.